### PR TITLE
Add APIs for adding dynamic files to IWorkspaceProjectContext

### PR DESCRIPTION
**Pull request by heejaechang**
@jasonmalinowski @rynowak here is one we discussed today for project system.

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs [hunk 1]
```diff
@@ -36,6 +38,8 @@ internal interface IWorkspaceProjectContext : IDisposable
         void RemoveSourceFile(string filePath);
         void AddAdditionalFile(string filePath, bool isInCurrentContext = true);
         void RemoveAdditionalFile(string filePath);
+        Task AddNonSourceFileAsync(string filePath, IEnumerable<string> folderNames = null, CancellationToken cancellationToken = default);
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs]: so what name should I use instead of NonSource?

Comment (rynowak) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs]: Dynamic/Ephemeral/Generated ?

Comment (rynowak) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs]: All 4? `AddDynamicNonSourceGeneratedEphemeralSourceFileAsync` :trollface:

**Review by rynowak — APPROVED** [review]
I gave my two cents on the name, the changes in the PR are what I was asking for 😁

**Review by rynowak — CHANGES_REQUESTED** [review]
Wait, commented too fast. Why is this async? We agreed yesterday that this needs to be sync to start

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs [hunk 2]
```diff
@@ -36,6 +38,8 @@ internal interface IWorkspaceProjectContext : IDisposable
...
+        Task RemoveNonSourceFileAsync(string fullPath, CancellationToken cancellationToken = default);
```
Comment (rynowak) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs]: These need to be sync for now.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs (hunk 1 above)
Comment (heejaechang) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs]: :)  Dynamic sounds good to me?

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 using System;
 using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
```
Comment (rynowak) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs:6]: Usings aren't needed now

**Review by rynowak — APPROVED** [review]

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs [hunk 4]
```diff
@@ -36,6 +38,8 @@ internal interface IWorkspaceProjectContext : IDisposable
         void RemoveSourceFile(string filePath);
         void AddAdditionalFile(string filePath, bool isInCurrentContext = true);
         void RemoveAdditionalFile(string filePath);
+        void AddDynamicFile(string filePath, IEnumerable<string> folderNames = null);
+        void RemoveDynamicFile(string fullPath);
```
Comment (rynowak) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs:42]: /cc @davkean

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs (hunk 3 above)
Comment (heejaechang) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs:6]: let me remove those on the main PR in FarRazor2

**Review by jasonmalinowski — APPROVED** [review]
Looks good, @heejaechang I'll hit the merge button when I'm ready to take it. Let me know if there's anything else you want done first.

**Pull request merged**